### PR TITLE
IFP: Multi height cell support

### PR DESCRIPTION
**Pull request by mina1460**
This PR would serve as the foundation for supporting multi-height cells.

It also includes support for removing the need for the -site name flag for init_floorplan command. It would still work with and without it.

### CODE REVIEW

File: src/ifp/include/ifp/InitFloorplan.hh [hunk 1]
```diff
@@ -35,8 +35,9 @@
 
 #pragma once
 
+#include <set>
 #include <string>
-
+#include <vector>
```
Comment (maliberty) [inline src/ifp/include/ifp/InitFloorplan.hh:40]: I don't see a usage of vector in this header.  It is needed elsewhere please add it to the file(s) that use it.

File: src/ifp/src/InitFloorplan.cc [hunk 2]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+{
+  std::set<dbSite*> sites;
+
+  // loop over all instantiated cells in the block
+  for (dbInst* inst : block_->getInsts()) {
+    dbMaster* master = inst->getMaster();
+    // get the site of the master
+    auto site = master->getSite();
+    if (site == nullptr) {
+      continue;
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:402]: A warning would be good here as this would be unexpected.

File: src/ifp/src/InitFloorplan.cc [hunk 3]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+  });
+
+  // smallest site height
+  double min_site_height = site_vec[0]->getHeight();
+  double factor = 1.0;
+  for (auto site : site_vec) {
+    double site_height = site->getHeight();
+    factor = site_height / min_site_height;
+    // check if the site height is a multiple of the smallest site height
+    if (std::ceil(factor) != std::floor(factor)) {
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:419]: Use integers here not doubles.  Check with the modulo operator not ceil/floor.

File: src/ifp/src/InitFloorplan.cc [hunk 4]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+      logger_->error(
+          IFP,
+          228,  // TODO: add error code
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:422]: Why is this a TODO?

File: src/ifp/src/InitFloorplan.cc [hunk 5]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+          "site height {} is not a multiple of the smallest site height {}",
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:423]: Give the site names too

File: src/ifp/src/InitFloorplan.cc [hunk 6]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+          site_height,
+          min_site_height);
+      // erase from the set if the site height is not a multiple of the smallest
+      sites.erase(site);
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:427]: you'll never get here as error() throws an exception.  The erase wouldn't be safe while iterating anyways

File: src/ifp/src/InitFloorplan.cc [hunk 7]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+    }
+  }
+  return sites;
+}
+
 void InitFloorplan::makeRows(dbSite* site,
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:433]: Please add a unit test that inserts double height rows.

File: src/ifp/src/InitFloorplan.cc [hunk 8]
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
...
+    sites.insert(site);
+  }
+
+  std::vector<dbSite*> site_vec{sites.begin(), sites.end()};
+  sort(site_vec.begin(), site_vec.end(), [](dbSite* a, dbSite* b) {
+    return a->getHeight() < b->getHeight();
+  });
+
+  // smallest site height
+  double min_site_height = site_vec[0]->getHeight();
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:413]: site_vec[0] is invalid if sites.empty()

File: src/ifp/src/InitFloorplan.cc [hunk 9]
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
...
+    dbSite* site_found = findSite(site_name.c_str());
+    if (site_found != nullptr) {
+      sites.insert(site_found);
+    }
+  }
+  // if the sites set is empty, then there are no sites in the design
+  // and user didn't specify a site that we could find
+  if (sites.empty()) {
+    logger_->error(IFP, 33, "No sites found.");
+    return;
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:188]: rm, not reached

File: src/ifp/src/InitFloorplan.cc [hunk 10]
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
...
+  }
+  // get min site height using std::min_element
+  int min_site_height
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:191]: const

File: src/ifp/src/InitFloorplan.cc [hunk 11]
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
...
+      = (*std::min_element(sites.begin(),
+                           sites.end(),
+                           [](dbSite* a, dbSite* b) {
+                             return a->getHeight() < b->getHeight();
+                           }))
+            ->getHeight();
+
+  eval_upf(logger_, block_);
+  for (auto site : sites) {
+    if (core.xMin() >= 0 && core.yMin() >= 0) {
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:201]: The core size won't change with sites so this check should be hoisted.

File: src/ifp/src/InitFloorplan.cc [hunk 12]
```diff
@@ -375,8 +446,9 @@ void InitFloorplan::makeRows(dbSite* site,
 
   int y = core_ly;
   for (int row = 0; row < rows_y; row++) {
-    dbOrientType orient = (row % 2 == 0) ? dbOrientType::R0   // N
-                                         : dbOrientType::MX;  // FS
+    dbOrientType orient = (factor == 2 or row % 2 == 0)
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:449]: any even factor not just 2 (eg quad height cells)

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/ifp/src/InitFloorplan.cc (hunk 4 above)
Comment (mina1460) [reply src/ifp/src/InitFloorplan.cc:422]: I didn't know how I should pick an error code, I thought I should document the code somewhere as each code seems to be unique for a message. Should I just pick the first available number?

Comment (maliberty) [reply src/ifp/src/InitFloorplan.cc:422]: Use the next available from ifp/messages.txt (ie 40)

File: src/ifp/src/InitFloorplan.cc (hunk 7 above)
Comment (maliberty) [reply src/ifp/src/InitFloorplan.cc:433]: This still needs addressing

**Comment by QuantamHD** [comment]
@mina1460 Would you mind rebasing to master I introduced a clang-tidy action that leaves comments on your PR that I would like to see here https://github.com/The-OpenROAD-Project/OpenROAD/pull/2980

**Comment by mina1460** [comment]
@QuantamHD Done

**Comment by QuantamHD** [comment]
Thanks!

File: src/ifp/test/mock.tcl [hunk 13]
```diff
@@ -0,0 +1,7 @@
+source "helpers.tcl"
+read_lef Nangate45/Nangate45.lef
+read_lef Nangate45/fake_macros.lef
+read_def mock.def
+initialize_floorplan -utilization 30 \
+  -aspect_ratio 0.5 \
+  -core_space 1
```
Comment (maliberty) [inline src/ifp/test/mock.tcl:7]: The test needs to do something to validate the results.  Likely you want a write_def and a .defok file.  Please see other tests for examples.

Comment (mina1460) [reply src/ifp/test/mock.tcl:7]: yes, absolutely. I just wanted to verify that the lef/def files and the mock macros are OK and this is indeed testing a double height cell before spending time on a defok file.

Should I proceed with it?

Comment (maliberty) [reply src/ifp/test/mock.tcl:7]: yes

File: src/ifp/test/mock.def [hunk 14]
```diff
@@ -5,7 +5,7 @@ DESIGN simple_design ;
 UNITS DISTANCE MICRONS 1000 ;
 DIEAREA ( 0 0 ) ( 2000 2000 ) ;
 ROW ROW_0 SingleHeightSite 0 0 FS DO 10 BY 1 STEP 1000 0 ;
-ROW ROW_1 DoubleHeightSite 1200 2200 N DO 12 BY 1 STEP 1100 0 ;
+ROW ROW_1 DoubleHeightSite 0 1000 N DO 10 BY 1 STEP 1000 0 ;
```
Comment (maliberty) [inline src/ifp/test/mock.def]: Why are there rows in the input DEF?  The goal of ifp is to build the rows.  You could just read a .v for this test which would better mirror ORFS (ie the yosys output).

Comment (mina1460) [reply src/ifp/test/mock.def]: makes sense. I will fix it

**Comment by QuantamHD** [comment]
@maliberty @vvbandeira I think @mina1460 is getting a high number of notifications from the sync upstream, because of the waiting for approval status. We should probably add him to the list of approved workflow runners.

**Comment by maliberty** [comment]
@QuantamHD  once his first PR is merged he will automatically be added.  I've never had to do it manually.

**Comment by maliberty** [comment]
The CI failures need to be addressed before this can be merged

**Comment by mina1460** [comment]
Hi @maliberty 

All CI/CD failures are now resolved. 

I modified the updateVoltageDomain and removed the site from it so it is no longer needed. As you said, this had the potential of cutting double-height rows into single-height rows, but I don't think this is a problem with this modification anymore.

**Review by maliberty — APPROVED** [review]

**Pull request merged**